api/lib/flight_job_script_api/system_command.rb

```diff
@@ -187,75 +182,151 @@ def run(&block)
+        # Get the exitstatus and prevent a busy loop
+        unless exitstatus
+          begin
+            _, status = Timeout.timeout(step) do
+              begin
+                Process.wait2(pid)
+              rescue Errno::ECHILD, Errno::ESRCH
+                # NOOP - See kill below
+              end
+            end
+            self.exitstatus = process_status(status) if status
+          rescue Timeout::Error
+            # NOOP
+          end
+        end
 
-        # If all else fails, say 128
-        else
-          FlightJobScriptAPI.logger.error "No exit code provided (pid: #{pid})!"
-          self.exitstatus = 128
+        # Read the data from stdout/stderr
+        loop_stdout = read_io(out_read, self.stdout) if loop_stdout
+        loop_stderr = read_io(err_read, self.stderr) if loop_stderr
+
+        # Kill the process after a timeout
+        unless exitstatus
+          now = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+          if now - start > FlightJobScriptAPI.config.command_timeout
+            FlightJobScriptAPI.logger.error("Sending #{signal} to: #{pid}")
+            begin
+              Process.kill(-Signal.list[signal], pid)
+            rescue Errno::ECHILD, Errno::ESRCH
+              # Rapidly starting SystemCommands may cause the fork to fail
+              FlightJobScriptAPI.logger.error(<<~ERROR.chomp)
+                Failed to start/locate process: #{pid}
+              ERROR
+              self.exitstatus = 128
+              return
+            end
+            signal = 'KILL'
+            start = now
+          end
         end
+      end
+    end
 
-        self.stdout = out_read.read
-        self.stderr = err_read.read
-
-        FlightJobScriptAPI.logger.debug <<~DEBUG
-
-          COMMAND: #{cmd.join(' ')}
-          USER: #{user}
-          PID: #{pid}
-          STATUS: #{exitstatus}
-          STDIN:
-          #{stdin.to_s}
-          STDOUT:
-          #{stdout}
-          STDERR:
-          #{stderr}
-        DEBUG
+    def process_status(status)
+      if status.exitstatus
+        status.exitstatus
+      elsif status.signaled?
+        FlightJobScriptAPI.logger.warn <<~WARN.chomp
+          Inferring exit code from signal #{Signal.signame(status.termsig)} (pid: #{pid})
+        WARN
+        status.termsig + 128
+      else
+        FlightJobScriptAPI.logger.error "No exit code provided (pid: #{pid})!"
+        128
       end
-    ensure
-      out_read.close unless out_read.closed?
-      out_write.close unless out_write.closed?
-      err_read.close unless err_read.closed?
-      err_write.close unless err_write.closed?
-      in_read.close unless in_read.closed?
-      in_write.close unless in_write.closed?
     end
 
-    private
+    def read_io(io, buffer)
+      begin
+        loop { buffer << io.read_nonblock(1024) }
+      rescue Errno::EWOULDBLOCK, Errno::EINTR, Errno::EAGAIN, IO::WaitReadable
+        # There *may* be more data
+        return true
```

Comment:
In this situation `IO::Select` is not required to prevent a "busy wait". The `Process.wait` prevents this regardless.

I'm going to review your other comments, and *may* make the change anyway.